### PR TITLE
fix the bug of incorrect status if there is error during the doDDL on some of the hosts

### PR DESCRIPTION
**Pull request by hzhzhang**
…tion.

PR checklist:
- Did you run ClangFormat ?
- Did you run CheckStyle ?
- Did you check the comment / log / exception conventions in Engineering code process wiki page ?
- Did you import unnecessary headers ?
- Did you surround `Daisy : starts/ends` for new code in existing ClickHouse code base ?

Please write user-readable short description of the changes:

### CODE REVIEW

File: src/DistributedMetadata/DDLService.cpp [hunk 1]
```diff
@@ -245,11 +245,17 @@ Int32 DDLService::doDDL(
         auto [response, http_code] = sendRequest(uri, method, query_id, user, password, payload, log);
 
         err = toErrorCode(http_code, response);
-        if (err == ErrorCodes::OK || err == ErrorCodes::UNRETRIABLE_ERROR)
+        if (err == ErrorCodes::OK)
         {
             return err;
         }
 
+        if (err == ErrorCodes::UNRETRIABLE_ERROR)
```
Comment (zzbdatatom) [inline src/DistributedMetadata/DDLService.cpp:253]: else if ?

File: src/DistributedMetadata/DDLService.cpp [hunk 2]
```diff
@@ -313,16 +320,23 @@ void DDLService::createTable(DWAL::RecordPtr record)
                 uri.addQueryParameter("distributed_ddl", "false");
                 uri.addQueryParameter("shard", std::to_string(j));
 
-                auto err = doDDL(payload, target_hosts[i * shards + j], Poco::Net::HTTPRequest::HTTP_POST, query_id, user);
-                if (err == ErrorCodes::UNRETRIABLE_ERROR)
+                auto err = doDDL(payload, uri, Poco::Net::HTTPRequest::HTTP_POST, query_id, user);
```
Comment (zzbdatatom) [inline src/DistributedMetadata/DDLService.cpp]: Let's use const auto & err

File: src/DistributedMetadata/DDLService.cpp [hunk 3]
```diff
@@ -413,13 +427,25 @@ void DDLService::mutateTable(DWAL::RecordPtr record, const String & method) cons
         return;
     }
 
+    std::vector<String> failed_hosts;
     for (auto & uri : target_hosts)
     {
         uri.setQueryParameters(Poco::URI::QueryParameters{{"distributed_ddl", "false"}});
-        doDDL(payload, uri, method, query_id, user);
+        auto err = doDDL(payload, uri, method, query_id, user);
```
Comment (zzbdatatom) [inline src/DistributedMetadata/DDLService.cpp]: 1. as line 323 .
2. Can we extract the same logic ?

File: src/DistributedMetadata/DDLService.cpp [hunk 4]
```diff
@@ -478,11 +504,25 @@ void DDLService::mutateDatabase(DWAL::RecordPtr record, const String & method) c
...
+        if (err != ErrorCodes::OK)
+        {
+            failed_hosts.push_back(uri.getHost() + ":" + toString(uri.getPort()));
+        }
+    }
+
+    if (failed_hosts.empty())
+    {
+        succeedDDL(query_id, user, payload);
```
Comment (zzbdatatom) [inline src/DistributedMetadata/DDLService.cpp]: Whether there is a logical conflict with line 528 ？

File: src/DistributedMetadata/DDLService.cpp (hunk 1 above)
Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp:253]: `else` is not need here as it  has already called `return` in the `if` above

File: src/DistributedMetadata/DDLService.cpp (hunk 4 above)
Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp]: good  catch.

File: src/DistributedMetadata/DDLService.cpp (hunk 3 above)
Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp]: fixed

File: src/DistributedMetadata/DDLService.cpp [hunk 5]
```diff
@@ -260,6 +266,31 @@ Int32 DDLService::doDDL(
     return err;
 }
 
+void DDLService::doDDLOnHosts(const std::vector<Poco::URI> & target_hosts, const String & payload,
+                              const String & method, const String & query_id, const String & user) const
+{
+    std::vector<String> failed_hosts;
+    /// FIXME : Parallelize doDDL on the uris
+    for (auto uri : target_hosts)
```
Comment (chenziliang) [inline src/DistributedMetadata/DDLService.cpp]: `for (auto & uri : target_hosts)` ?

File: src/DistributedMetadata/DDLService.cpp [hunk 6]
```diff
@@ -260,6 +266,31 @@ Int32 DDLService::doDDL(
...
+    {
+        uri.addQueryParameter("distributed_ddl", "false");
+        const auto & err = doDDL(payload, uri, method, query_id, user);
```
Comment (chenziliang) [inline src/DistributedMetadata/DDLService.cpp]: `auto err = doDDL(...)`

File: src/DistributedMetadata/DDLService.cpp [hunk 7]
```diff
@@ -260,6 +266,31 @@ Int32 DDLService::doDDL(
...
+        if (err != ErrorCodes::OK)
+        {
+            failed_hosts.push_back(uri.getHost() + ":" + toString(uri.getPort()));
```
Comment (chenziliang) [inline src/DistributedMetadata/DDLService.cpp:280]: `uri.toString()` ?

Comment (chenziliang) [reply src/DistributedMetadata/DDLService.cpp:280]: do we need break the loop if we encounter error or we continue to execute the rest request to other hosts ?

Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp:280]: 1. uri.toString() will print the full url.
2. Currently, we will continue to execute the rest request to other hosts. We can modify the logic when we import rollback in future.

File: src/DistributedMetadata/DDLService.cpp (hunk 5 above)
Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp]: fixed

File: src/DistributedMetadata/DDLService.cpp (hunk 6 above)
Comment (hzhzhang) [reply src/DistributedMetadata/DDLService.cpp]: fixed

File: src/DistributedMetadata/DDLService.cpp (hunk 7 above)
Comment (chenziliang) [reply src/DistributedMetadata/DDLService.cpp:280]: What's the point of continuing the execution for now ? Can we just fail the operation now ?

Comment (chenziliang) [reply src/DistributedMetadata/DDLService.cpp:280]: We can fix this behavior in a different PR

**Pull request merged**